core/src/main/java/org/apache/calcite/plan/volcano/RelSubset.java

```diff
@@ -429,8 +471,48 @@ public RelNode visit(
           // out why we reached impasse.
           StringWriter sw = new StringWriter();
           final PrintWriter pw = new PrintWriter(sw);
-          pw.println("Node [" + subset.getDescription()
-              + "] could not be implemented; planner state:\n");
+
+          DeadEndFinder finder = new DeadEndFinder();
+          finder.visit(subset);
+          pw.println("Node ");
+          pw.print(subset.getDescription());
+          pw.print(" is not implementable. It means there's not enough rules to produce the node"
+              + " with desired properties");
```

Comment:
I would suggest wording the last statement as "There are not enough rules to produce a node with the desired properties."